### PR TITLE
All: Working MSVC support. 

### PR DESCRIPTION
**Pull request by waddlesplash**

### CODE REVIEW

File: src/platform/qt/InputProfile.h [hunk 1]
```diff
@@ -48,15 +48,9 @@ class InputProfile {
...
-	                       AxisValue axes[GBA_KEY_MAX] = (AxisValue[GBA_KEY_MAX]) {
-	                       	                             {}, {}, {}, {}, 
-	                                                     { GamepadAxisEvent::Direction::POSITIVE, 0 },
-	                                                     { GamepadAxisEvent::Direction::NEGATIVE, 0 },
-	                                                     { GamepadAxisEvent::Direction::NEGATIVE, 1 },
-	                                                     { GamepadAxisEvent::Direction::POSITIVE, 1 },
-	                                                     {}, {}},
+						   const ShortcutButton* shortcutButtons,
+						   const ShortcutAxis* shortcutAxes,
+						   AxisValue axes[GBA_KEY_MAX],
```
Comment (waddlesplash) [inline src/platform/qt/InputProfile.h]: I took this out because I was getting
> error C4576: a parenthesized type followed by an initializer list is a non-standard explicit type conversion syntax

File: src/platform/qt/InputProfile.cpp [hunk 2]
```diff
@@ -154,6 +155,7 @@ const InputProfile InputProfile::s_defaultMaps[] = {
 		}
 	},
 };
+#endif
```
Comment (waddlesplash) [inline src/platform/qt/InputProfile.cpp]: Same C4576 here

File: src/platform/qt/OverrideView.cpp [hunk 3]
```diff
@@ -56,12 +56,12 @@ void OverrideView::saveOverride() {
...
 
 void OverrideView::updateOverrides() {
-	m_override = (GBACartridgeOverride) {
+	/*m_override = (GBACartridgeOverride) {
 		"",
 		static_cast<SavedataType>(m_ui.savetype->currentIndex() - 1),
 		HW_NO_OVERRIDE,
 		IDLE_LOOP_NONE
-	};
+	};*/
```
Comment (waddlesplash) [inline src/platform/qt/OverrideView.cpp]: And here

File: src/util/nointro.c [hunk 4]
```diff
@@ -158,7 +158,8 @@ struct NoIntroDB* NoIntroDBLoad(struct VFile* vf) {
 	struct {
 		char* key;
 		struct NoIntroItem* item;
-	} keyStack[KEY_STACK_SIZE] = {};
+	} keyStack[KEY_STACK_SIZE];
+	//keyStack = {{}};
```
Comment (waddlesplash) [inline src/util/nointro.c]: I think this was a different error but now I can't recall

File: src/platform/msvc/dirent.h [hunk 5]
```diff
@@ -0,0 +1,917 @@
+/*
```
Comment (endrift) [inline src/platform/msvc/dirent.h]: Please write a VDir for Win32 instead of using a dirent shim. The whole point of VDir is to have platform-specific implementations instead of having to reimplement all of dirent each time.

File: src/platform/qt/PaletteView.h [hunk 6]
```diff
@@ -13,8 +13,6 @@
 
 #include "ui_PaletteView.h"
 
-class GameController;
```
Comment (endrift) [inline src/platform/qt/PaletteView.h]: Why?

Comment (waddlesplash) [inline src/platform/qt/PaletteView.h]: Uh, you include GameController.h on line 11 so this is redundant and it causes a compiler error?

**Comment by endrift** [comment]
Please rebase on top of 3a51c59768a0e4bbba9dc516b0554eb520474c05; it adds native VDir stuff so you don't need to worry about the dirent shim anymore.

**Comment by waddlesplash** [comment]
Sorry, this week has been madness for me. It looks like you already merged most of this? I'll get the rest over the weekend.

**Comment by endrift** [comment]
I only merged the removal of the forward declaration. Everything else that would be a conflicting change I did somewhat differently, e.g. I do actually include Windows.h in common.h now. Also, I wrote the Win32 VDir so you can remove the dirent shim entirely.

**Comment by waddlesplash** [comment]
As soon as https://github.com/mgba-emu/mgba/blob/master/src/platform/qt/OverrideView.cpp#L59 is switched off of C99 syntax, this will make mGBA master work with MSVC 100%.

File: CMakeLists.txt [hunk 7]
```diff
@@ -147,6 +147,10 @@ if(WIN32)
 	list(APPEND CORE_VFS_SRC ${CMAKE_SOURCE_DIR}/src/util/vfs/vfs-fd.c ${CMAKE_SOURCE_DIR}/src/platform/windows/vfs-w32.c)
 	file(GLOB OS_SRC ${CMAKE_SOURCE_DIR}/src/platform/windows/*.c)
 	source_group("Windows-specific code" FILES ${OS_SRC})
+	if(MSVC)
+		include_directories(SYSTEM ${CMAKE_SOURCE_DIR}/src/platform/msvc)
```
Comment (endrift) [inline CMakeLists.txt]: Why did you make a new folder for this, when platform/windows already exists? getopt.h is only included in one place, so there's no need to add it to the include path either as specializing that include should work.

**Comment by endrift** [comment]
Removing that bit of C99ism could be part of this PR. Just remove the struct name and parens and it should work.

File: CMakeLists.txt (hunk 7 above)
Comment (waddlesplash) [inline CMakeLists.txt]: Because this is only for MSVC. MinGW and friends have their own getopt.h, and I didn't even test with those; for all I can tell this getopt was designed for and only for MSVC.

Comment (waddlesplash) [inline CMakeLists.txt]: I suppose I can just put getopt in windows as "getopt_msvc" and add some ifdefs.

File: src/platform/commandline.c [hunk 8]
```diff
@@ -20,7 +20,11 @@
 #include "util/string.h"
 
 #include <fcntl.h>
+#ifdef _MSC_VER
+#include "platform/windows/getopt_msvc.h"
```
Comment (endrift) [inline src/platform/commandline.c]: Why did you rename the file? platform/windows isn't in the search path for this file, so it shouldn't matter that the names collide.

File: src/platform/windows/getopt_msvc.h [hunk 9]
```diff
@@ -0,0 +1,607 @@
+#ifndef __GETOPT_H__
```
Comment (endrift) [inline src/platform/windows/getopt_msvc.h]: Please add a mention of this file to the copyright section of the README.
